### ISSUE TITLE
Virtual Environments - Install & Run HSP2 Python on Windows

### ISSUE THREAD

**Issue by jdkleiner** [HARPgroup/HSPsquared#55]
TL;DR This method is preferred for running HSP2 on windows because it keeps hsp2 and all of its dependencies installed in a virtual environment using its own python version and package versions. This method is also preferred because it prevents the user from having to make any tweaks to the file `HSPsquared/HSP2/environment.yml` in order to run hsp2 https://github.com/HARPgroup/HSPsquared/issues/48

Note: I'm still annotating some of the steps, but its actually pretty straightforward once you dive in 

0) install python version of interest
   - installation executables of previous versions of python can be found here: https://www.python.org/ftp/python/3.8.9/
   - I installed version 3.8.9 as HSP2 requires python =3.8*
      - For the installer, use `python-3.8.9-amd64.exe`
1) create a local folder for housing virtual environments, I went with: "C:\Users\nrf46657\virtualenvs"
2) open a PS terminal within that directory 
   `PS C:\Users\nrf46657\virtualenvs>`
3) create a virtual environment, utilizing a specific install of python (i.e. the 3.8 version installed above)
   ```
   PS C:\Users\nrf46657\virtualenvs> C:\Users\nrf46657\AppData\Local\Programs\Python\Python38\python.exe -m virtualenv hsp2dev_venv
   created virtual environment CPython3.8.9.final.0-64 in 1125ms
     creator CPython3Windows(dest=C:\Users\nrf46657\virtualenvs\hsp2dev_venv, clear=False, no_vcs_ignore=False, global=False)
     seeder FromAppData(download=False, pip=bundle, setuptools=bundle, wheel=bundle, via=copy, app_data_dir=C:\Users\nrf46657\AppData\Local\pypa\virtualenv)
       added seed packages: pip==22.3.1, setuptools==65.6.3, wheel==0.38.4
     activators BashActivator,BatchActivator,FishActivator,NushellActivator,PowerShellActivator,PythonActivator
   ```
   - Take a look at the packages in the virtual environment 
   ```
   PS C:\Users\nrf46657\virtualenvs> ls hsp2dev_venv\Lib\site-packages
       Directory: C:\Users\nrf46657\virtualenvs\hsp2dev_venv\Lib\site-packages


   Mode                 LastWriteTime         Length Name
   ----                 -------------         ------ ----
   d-----         1/10/2023   4:04 PM                pip
   d-----         1/10/2023   4:04 PM                pip-22.3.1.dist-info
   d-----         1/10/2023   4:04 PM                pkg_resources
   d-----         1/10/2023   4:04 PM                setuptools
   d-----         1/10/2023   4:04 PM                setuptools-65.6.3.dist-info
   d-----         1/10/2023   4:04 PM                wheel
   d-----         1/10/2023   4:04 PM                wheel-0.38.4.dist-info
   d-----         1/10/2023   4:04 PM                _distutils_hack
   -a----         1/10/2023   4:04 PM            151 distutils-precedence.pth
   -a----         1/10/2023   4:04 PM              0 pip-22.3.1.virtualenv
   -a----         1/10/2023   4:04 PM              0 setuptools-65.6.3.virtualenv
   -a----         1/10/2023   4:04 PM              0 wheel-0.38.4.virtualenv
   -a----         1/10/2023   4:04 PM             18 _virtualenv.pth
   -a----         1/10/2023   4:04 PM           5770 _virtualenv.py
   ```
4) Activate a virtual environment
   - First open a Command Prompt cmd
   - Navigate to directory housing your virtual environments 
   ```
   R:\>cd /d C:\Users\nrf46657\virtualenvs
   C:\Users\nrf46657\virtualenvs>
   ```
   - **hsp2dev_venv\Scripts\activate**
   ```
   C:\Users\nrf46657\virtualenvs>hsp2dev_venv\Scripts\activate

   (hsp2dev_venv) C:\Users\nrf46657\virtualenvs>
   ```
   - You can verify the version of Python being used in the virtual environment 
   ```
   (hsp2dev_venv) C:\Users\nrf46657\Desktop\GitHub\HSPsquared>python -V
   Python 3.8.9
   ```
5) install hsp2 within that virtual environment
   - still within cmd: change directory to HSPsquared
   ```
   (hsp2dev_venv) C:\Users\nrf46657\virtualenvs>cd 
   C:\Users\nrf46657\Desktop\GitHub\HSPsquared

   (hsp2dev_venv) C:\Users\nrf46657\Desktop\GitHub\HSPsquared>
   ```
   - install hsp2 
   ``` 
   (hsp2dev_venv) C:\Users\nrf46657\Desktop\GitHub\HSPsquared>pip install .
   Processing c:\users\nrf46657\desktop\github\hspsquared
     Preparing metadata (setup.py) ... done
   Collecting scipy
     Using cached scipy-1.10.0-cp38-cp38-win_amd64.whl (42.2 MB)
   Collecting pandas==1.2.*
     Using cached pandas-1.2.5-cp38-cp38-win_amd64.whl (9.3 MB)
   Collecting numba==0.53.1
     Using cached numba-0.53.1-cp38-cp38-win_amd64.whl (2.3 MB)
   Collecting numpy==1.20.*
     Using cached numpy-1.20.3-cp38-cp38-win_amd64.whl (13.7 MB)
   ...
   ```
   - This is super powerful because a) hsp2 gets installed within the virtual environment you created AND b) all of its hsp2 and all of its dependencies with the correct versions get installed within the virtual environment as well (i.e. all of the packages in the environment.yml get installed)
6) Verify success in PowerShell
   - Just as before, take a look at the packages in the virtual environment (there should now be many more packages listed, including HSP2) 
      - `PS C:\Users\nrf46657\virtualenvs> ls hsp2dev_venv\Lib\site-packages`
7) Verify you can now run hsp2 (This is best done from within VS Code)
   ```
   PS C:\Users\nrf46657\Desktop\GitHub\HSPsquared\tests\testcbp\HSP2results>
   ```
   - To activate your virtual environment, you must select the python executable
      - F1 -> Python: Select Interpreter -> Enter Interpreter Path... -> "C:\Users\nrf46657\virtualenvs\hsp2dev_venv\Scripts\python.exe"
      - **I need to finished fleshing out these notes

**Comment by rburghol** [comment]
Sow version of python:
```
S C:\usr\local\home\git\HSPsquared> py.exe -V  
Python 3.10.8
PS C:\usr\local\home\git\HSPsquared> py -0p
Installed Pythons found by C:\Users\faw18626\AppData\Local\Programs\Python\Launcher\py.exe Launcher for Windows
 -3.10-64       C:\usr\local\bin\python\python.exe *
 -3.8-64        C:\Users\faw18626\AppData\Local\Programs\Python\Python38\python.exe

PS C:\usr\local\home\git\HSPsquared> 
```

**Comment by rburghol** [comment]
Activate with gitbash (not powershell)
```
cd "C:\Users\faw18626\virtualenvs"
source hsp2dev_venv/Scripts/activate

# otherwise works as above with powershell
```

**Comment by rburghol** [comment]
Run hsp2 in interactive session in vscode
- install PyQt5 (to allow running interactive mode): `python -m pip install PyQt5 --trusted-host pypi.org --trusted-host pypi.python.org --trusted-host files.pythonhosted.org`

**Comment by rburghol** [comment]
#### Setup p10 on Windows in PowerShell
```
# install the necessary virtual environment modules
C:\usr\local\bin\python\python.exe  -m pip install --trusted-host pypi.org --trusted-host pypi.python.org --trusted-host files.pythonhosted.org --user virtualenv --default-timeout=100
C:\usr\local\bin\python\python.exe  -m pip install --trusted-host pypi.org --trusted-host pypi.python.org --trusted-host files.pythonhosted.org --default-timeout=100 pyreadline3

cd "C:\Users\faw18626\virtualenvs"
C:\usr\local\bin\python\python.exe -m virtualenv hsp2dev_py10
# Activate it
./hsp2dev_py10/Scripts/activate

cd c:/usr/local/home/git/HSPsquared
pip install --trusted-host pypi.org --trusted-host pypi.python.org --trusted-host files.pythonhosted.org --default-timeout=100 -e . 

```
#### Activate it for use in gitbash shell with py and hsp2
```
cd "C:\Users\faw18626\virtualenvs"

# Actviate in Bash
source ./hsp2dev_py10/Scripts/activate

cd c:/usr/local/home/git/HSPsquared

```

#### Activate it for use in PowerShell with py and hsp2
```
cd "C:\Users\faw18626\virtualenvs"

# Activate it in PS shell:
./hsp2dev_py10/Scripts/activate

cd c:/usr/local/home/git/HSPsquared

```

**Comment by rburghol** [comment]
Run and disable numba, which allows to do without a recompile:
- `NUMBA_DISABLE_JIT=1 hsp2 run test10specl.h5`

**Comment by rburghol** [comment]
Need to get venv going on linux... I suppose this should be as easy as activating the venv when running the meta model?  ANyhow, for now I did;
```
sudo python3.9 -m pip install --upgrade pip
sudo pip3.10 install .


```

**Comment by rburghol** [comment]
moved to https://github.com/HARPgroup/HSPsquared/issues/108
